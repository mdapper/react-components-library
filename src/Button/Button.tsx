import React, { ReactNode, MouseEvent } from 'react';
import styled from 'styled-components';
import { darken } from 'polished';

interface Props {
  children: ReactNode;

  /**
   * Disables the button, changing the visual style and make it unable to be pressed
   * @default false
   */
  disabled?: boolean;

  /**
   * Loading state when waiting for an action to complete
   * @default false
   */
  loading?: boolean;

  /**
   * The type of the button
   * @default button
   */
  type?: 'submit' | 'button';

  /** Function to handle button click */
  onClick?(e: MouseEvent<HTMLElement>): void;
}

interface StyledProps {
  disabled: boolean;
  loading?: boolean;
}

export const Button = ({
  children,
  disabled = false,
  loading = false,
  onClick,
  type = 'button',
}: Props) => (
  <ButtonStyled
    data-testid="button"
    disabled={disabled}
    loading={loading}
    onClick={onClick}
    type={type}
  >
    {children}
  </ButtonStyled>
);

const ButtonStyled = styled.button<StyledProps>`
  align-items: center;
  background-color: #9575cd;
  border-radius: 4px;
  border: 1px solid #9575cd;
  color: #fff;
  cursor: ${(props) => (props.disabled ? 'not-allowed' : 'pointer')};
  display: flex;
  font-size: 16px;
  font-weight: bold;
  height: 40px;
  justify-content: center;
  margin: 0;
  min-width: 140px;
  opacity: ${(props) => (props.disabled ? 0.6 : 1)};
  overflow: hidden;
  padding: 0 16px;
  transition: border-color 300ms ease, background-color 300ms ease,
    color 300ms ease;

  &:hover {
    background-color: ${darken(0.1, '#9575cd')};
    border-color: ${darken(0.1, '#9575cd')};
  }

  &:focus,
  &:active {
    background-color: ${darken(0.2, '#9575cd')};
    border-color: ${darken(0.2, '#9575cd')};
  }
`;
