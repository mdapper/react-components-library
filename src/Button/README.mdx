---
name: Button
route: /Button
menu: Components
---

import { Playground, Props } from 'docz';
import { Button } from './';

# Button

Basic Button component using Styled Components

## Usage

<Playground>
  <Button>Click here</Button>
</Playground>

## Props

<Props of={Button} />
