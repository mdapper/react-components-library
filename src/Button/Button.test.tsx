import * as React from 'react';
import { render } from '@testing-library/react';

import { Button } from './Button';

describe('<Button />', () => {
  it('renders', () => {
    const { container } = render(<Button>Click here</Button>);

    expect(container.firstChild).toMatchSnapshot();
  });
});
