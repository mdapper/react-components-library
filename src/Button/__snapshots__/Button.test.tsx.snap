// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #9575cd;
  border-radius: 4px;
  border: 1px solid #9575cd;
  color: #fff;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 16px;
  font-weight: bold;
  height: 40px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0;
  min-width: 140px;
  opacity: 1;
  overflow: hidden;
  padding: 0 16px;
  -webkit-transition: border-color 300ms ease,background-color 300ms ease, color 300ms ease;
  transition: border-color 300ms ease,background-color 300ms ease, color 300ms ease;
}

.c0:hover {
  background-color: #7850bf;
  border-color: #7850bf;
}

.c0:focus,
.c0:active {
  background-color: #603ba1;
  border-color: #603ba1;
}

<button
  class="c0"
  data-testid="button"
  type="button"
>
  Click here
</button>
`;
